src/chunkserver/chunkserver.cpp

```diff
@@ -75,8 +90,127 @@ DEFINE_string(walFilePoolDir, "./0/", "WAL filepool location");
                                     "WAL filepool meta path");
 
+DEFINE_string(pfs_cluster,
+    "",
+    "pfs cluster name, only for pfs");
+
+DEFINE_string(pfs_pbd_name,
+    "",
+    "pfs pbd name, only for pfs");
+DEFINE_int32(pfs_host_id,
+    -1,
+    "pfs host id, only for pfs");
+
 const char* kProtocalCurve = "curve";
 
+
+#ifdef WITH_SPDK
+void
+glog_pfs_func(int level, const char *file, const char *func, int line,
+    const char *fmt, va_list ap) {
+    char buf[8192];
+
+    int glevel = google::GLOG_INFO;
+    vsnprintf(buf, sizeof(buf), fmt, ap);
+    switch (level) {
+    case PFS_TRACE_FATAL:
+        glevel = google::GLOG_FATAL;
+        google::LogMessage(file, line, glevel).stream() << buf;
+        abort();
+        break;
+    case PFS_TRACE_ERROR:
+        glevel = google::GLOG_ERROR;
+        break;
+    case PFS_TRACE_WARN:
+        glevel = google::GLOG_WARNING;
+        break;
+    case PFS_TRACE_INFO:
+    case PFS_TRACE_DBG:
+    case PFS_TRACE_VERB:
+    default:
+        glevel = google::GLOG_INFO;
+        break;
+    }
+    google::LogMessage(file, line, glevel).stream() << buf;
+}
+
+void
+glog_spdk_func(int level, const char *a_file, const int a_line,
+    const char *func, const char *fmt, va_list ap) {
+    char buf[8192];
+    const char *file = a_file;
+    int line = a_line;
+
+    if (file == NULL) {
+        file = "<spdk>";
+        if (a_line <= 0)
+            line = 1;
+    }
+
+    int glevel = google::GLOG_INFO;
+    vsnprintf(buf, sizeof(buf), fmt, ap);
+    switch (level) {
+    case SPDK_LOG_ERROR:
+        glevel = google::GLOG_ERROR;
+        break;
+    case SPDK_LOG_WARN:
+        glevel = google::GLOG_WARNING;
+        break;
+    case SPDK_LOG_NOTICE:
+    case SPDK_LOG_INFO:
+        glevel = google::GLOG_INFO;
+        break;
+#ifndef NDEBUG
+    case SPDK_LOG_DEBUG:
+        glevel = google::GLOG_INFO;
+        break;
+#endif
+    }
+    google::LogMessage(file, line, glevel).stream() << buf;
+}
+
+static ssize_t
+glog_dpdk_log_func(void *cookie, const char *buf, size_t size) {
+    int level = rte_log_cur_msg_loglevel();
+    int glevel = google::GLOG_INFO;
+
+    switch (level) {
+    case RTE_LOG_EMERG:
+    case RTE_LOG_ALERT:
+    case RTE_LOG_CRIT:
+    case RTE_LOG_ERR:
+        glevel = google::GLOG_ERROR;
+        break;
+    case RTE_LOG_WARNING:
+        glevel = google::GLOG_WARNING;
+        break;
+    case RTE_LOG_NOTICE:
+    case RTE_LOG_INFO:
+    default:
+        glevel = google::GLOG_INFO;
+        break;
+#ifndef NDEBUG
+    case RTE_LOG_DEBUG:
+        glevel = google::GLOG_INFO;
+        return 0;
+#endif
+    }
+    google::LogMessage("<dpdk>", 0, glevel).stream().write(buf, size);
+    if (level == RTE_LOG_EMERG)
+        abort();
+    return size;
+}
+
+void* dpdk_mem_allocate(size_t align, size_t sz) {
+    /* rte_malloc seems fast enough, otherwise we need to use mempool */
+    return rte_malloc("iobuf", sz, align);
```

Comment:
Is rte_malloc a new memory allocation library?

Reply:
It is dpdk memory allocation